storage/innobase/sync/srw_lock.cc

```diff
@@ -392,14 +402,45 @@ template void ssux_lock_impl<false>::wr_wait(uint32_t);
 template<bool spinloop>
 void ssux_lock_impl<spinloop>::rd_wait()
 {
+  const unsigned delay= srw_pause_delay();
+
+  if (spinloop)
+  {
+    for (auto spin= srv_n_spin_wait_rounds; spin; spin--)
+    {
+      srw_pause(delay);
+      if (rd_lock_try())
+        return;
+    }
+  }
+
+  /* Subscribe to writer.wake() or write.wake_all() calls by
+  concurrently executing rd_wait() or writer.wr_unlock(). */
+  uint32_t wl= 1 + writer.lock.fetch_add(1, std::memory_order_acquire);
+
   for (;;)
   {
-    writer.wr_lock();
-    bool acquired= rd_lock_try();
-    writer.wr_unlock();
-    if (acquired)
+    if (UNIV_LIKELY(writer.HOLDER & wl))
+      writer.wait(wl);
+    uint32_t lk= rd_lock_try_low();
+    if (!lk)
       break;
+    if (UNIV_UNLIKELY(lk == WRITER)) /* A wr_lock() just succeeded. */
+      writer.wake_all(); /* Immediately wake up (also) wr_lock(). */
```

Comment:
There could be several of these `for (;;)` loops executing in multiple threads, each of them ‘consuming’ wake-ups in `wait()`. The `wr_unlock()` would only wake up one of these threads.

If we invoked `wake()` instead of `wake_all()`, then this wake-up signal could go into a waste, waking up another `rd_wait()` instead of waking the `wr_lock()` which is the only thread that can make progress. Only after the blocked `wr_lock()` thread does `wr_unlock()` can all the `rd_lock()` make progress.

The `wake_all()` call at the end I added ‘just in case’; initially it was `wake()`. I don’t remember if I observed a hang when I had a `wake()` at that spot.